File: nodejs/controllers/dsacontroller.js
const { QuestionRequest } = require('../models/dsaRequest');
const DsaService = require('../services/dsaService');
const OpenAIService = require('../services/openaiService');

class DsaController {
    constructor() {
        const openaiService = new OpenAIService('your_openai_api_key_here', 'gpt-3.5-turbo');
        this.dsaService = new DsaService(openaiService);
    }

    registerRoutes(app) {
        app.post('/analyze_dsa', this.analyzeDsa.bind(this));
    }

    async analyzeDsa(req, res) {
        try {
            const questionRequest = new QuestionRequest(req.body);
            const analysis = await this.dsaService.analyzeQuestion(questionRequest);
            res.json({ question: questionRequest.question, analysis });
        } catch (error) {
            res.status(500).json({ error: error.message });
        }
    }
}

module.exports = DsaController;

File: python/controllers/dsa_controller.py
from fastapi import APIRouter, HTTPException
from models.dsa_request import QuestionRequest
from services.dsa_service import DsaService
from services.openai_service import OpenAIService

class DsaController:
    def __init__(self):
        self.router = APIRouter()
        self.router.add_api_route("/analyze_dsa", self.analyze_dsa, methods=["POST"])
        openai_service = OpenAIService(api_key="your_openai_api_key_here", model="gpt-3.5-turbo")
        self.dsa_service = DsaService(openai_service=openai_service)

    async def analyze_dsa(self, question_request: QuestionRequest):
        try:
            analysis = self.dsa_service.analyze_question(question_request)
            return {"question": question_request.question, "analysis": analysis}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))
